Issues 38 and 39 (#41)
NTAP/rfc8312bis: draft-eggert-tcpm-rfc8312bis.md

--- abstract

CUBIC is an extension to the traditional TCP standards. It differs from
the traditional TCP standards only in the congestion control algorithm on
the sender side. In particular, it uses a cubic function instead of a
linear window increase function of the traditional TCP standards to
improve scalability and stability under fast and long-distance
networks. CUBIC has been adopted as

> the default TCP congestion control algorithm by Linux, Android,

Reviewer: Is Android not Linux?

Author: ok, I will remove it.

> Windows, and Apple stacks. This document
> provides a specification of CUBIC to enable third-party
> implementations and to solicit community feedback through
> experimentation on the performance of CUBIC.

Reviewer: ```suggestion
Windows, and Apple stacks. This document
updates the specification of CUBIC to include algorithmic
improvements based on these implementations and recent
academic work. Based on the extensive deployment experience
with CUBIC, it also moves the specification to the Standards Track.
```

problem is equally applicable to all Reno-style TCP standards and
their variants, including TCP-Reno {{!RFC5681}}, TCP-NewReno
{{!RFC6582}}{{!RFC6675}}, SCTP {{?RFC4960}}, and TFRC {{!RFC5348}},
which use the same linear increase function for window growth.
We refer to all Reno-style TCP standards and
their variants collectively as "AIMD TCP", below because they use
the Additive Increase and Multiplicative Decrease algorithm (AIMD).

CUBIC, originally proposed in {{HRX08}}, is a modification to the

> congestion control algorithm of AIMD TCP to remedy this problem.

Reviewer: ```suggestion
congestion control algorithm of traditional Additive Increase and
Multiplicative Decrease (AIMD) TCP to remedy this problem.
```

This document describes the most recent specification of CUBIC.
Specifically, CUBIC uses a cubic function instead of a linear window
increase function of AIMD TCP to improve scalability and stability
under fast and long-distance networks.

Binary Increase Congestion Control (BIC-TCP) {{XHR04}}, a predecessor
of CUBIC, was selected as the default TCP congestion control algorithm
by Linux in the year 2005 and had been used for several years by the
Internet community at large. CUBIC uses a similar window increase
function as BIC-TCP and is designed to be less aggressive and fairer
to AIMD TCP in bandwidth usage than BIC-TCP while maintaining the
strengths of BIC-TCP such as stability, window scalability, and RTT
fairness. CUBIC has been adopted as the default TCP

> congestion control algorithm in Linux, Andoid, Windows,

Reviewer: Same a comment about Android using Linux?

Author: removed

and Apple stacks, and has been used and deployed globally.

> Through extensive testing in various Internet scenarios, we
> believe that CUBIC is safe for deployment in the global
> Internet.

Reviewer: ```suggestion
Extensive, decade-long deployment experience in vastly different
Internet scenarios has convincingly demonstrated that CUBIC is
safe for deployment on the global Internet and delivers substantial
benefits over traditional AIMD congestion control. It is therefore to
be regarded as the current standard for TCP congestion control.
```

problem is equally applicable to all Reno-style TCP standards and
their variants, including TCP-Reno {{!RFC5681}}, TCP-NewReno
{{!RFC6582}}{{!RFC6675}}, SCTP {{?RFC4960}}, and TFRC {{!RFC5348}},
which use the same linear increase function for window growth.
We refer to all Reno-style TCP standards and

> their variants collectively as "AIMD TCP", below because they use

Author: done

> Principle 1: For better network utilization and stability, CUBIC

Reviewer: Would it be possible to turn the following four very long paragraphs on these different four principles into sub-sections? Could someone suggest for section headings to use?

Author: Done.

Note that once *W<sub>est</sub>* reaches *W<sub>max</sub>*, that is,
*W<sub>est</sub>* >= *W<sub>max</sub>*, {{{α}{}}}*<sub>aimd</sub>*
SHOULD be set to 1 to achieve the same congestion window increment as

> AIMD TCP that uses AIMD(1, 0.5).

Reviewer: ```suggestion
AIMD TCP, which uses AIMD(1, 0.5).
```

| 1.0e-08 | 12000 | 574356 | 18740 | 33325 | 59261 |
{: #tab2 title="AIMD TCP, HSTCP, and CUBIC with RTT = 0.01 seconds"}

{{tab2}} describes the response function of AIMD TCP, HSTCP, and
CUBIC in networks with *RTT* = 0.01 seconds. The average window size is
in MSS-sized segments.

Both tables show that CUBIC with any of these three *C* values is more
friendly to AIMD TCP than HSTCP, especially in networks with a short *RTT*
where AIMD TCP performs reasonably well. For example, in a network with
*RTT* = 0.01 seconds and p=10^-6, AIMD TCP has an average window of 1200
packets. If the packet size is 1500 bytes, then AIMD TCP can achieve an
average rate of 1.44 Gbps. In this case, CUBIC with *C*=0.04 or
*C*=0.4 achieves exactly the same rate as AIMD TCP, whereas HSTCP
is about ten times more aggressive than AIMD TCP.

We can see that *C* determines the aggressiveness of CUBIC in
competing with other congestion control algorithms for bandwidth.
CUBIC is more friendly to AIMD TCP, if the value of *C* is lower.
However, we do not recommend setting *C* to a very low value like
0.04, since CUBIC with a low *C* cannot efficiently use the bandwidth

> in fast and long-distance networks. Based on these observations

Reviewer: ```suggestion
in fast, long-distance networks. Based on these observations
```

> and our experiments, we find *C*=0.4 gives a good balance between AIMD-

Reviewer: ```suggestion
and extensive deployment experience, we find *C*=0.4 gives
a good balance between AIMD-
```

| 1.0e-08 | 12000 | 574356 | 18740 | 33325 | 59261 |
{: #tab2 title="AIMD TCP, HSTCP, and CUBIC with RTT = 0.01 seconds"}

{{tab2}} describes the response function of AIMD TCP, HSTCP, and
CUBIC in networks with *RTT* = 0.01 seconds. The average window size is
in MSS-sized segments.

Both tables show that CUBIC with any of these three *C* values is more
friendly to AIMD TCP than HSTCP, especially in networks with a short *RTT*
where AIMD TCP performs reasonably well. For example, in a network with
*RTT* = 0.01 seconds and p=10^-6, AIMD TCP has an average window of 1200
packets. If the packet size is 1500 bytes, then AIMD TCP can achieve an
average rate of 1.44 Gbps. In this case, CUBIC with *C*=0.04 or
*C*=0.4 achieves exactly the same rate as AIMD TCP, whereas HSTCP
is about ten times more aggressive than AIMD TCP.

We can see that *C* determines the aggressiveness of CUBIC in
competing with other congestion control algorithms for bandwidth.
CUBIC is more friendly to AIMD TCP, if the value of *C* is lower.
However, we do not recommend setting *C* to a very low value like
0.04, since CUBIC with a low *C* cannot efficiently use the bandwidth

> in fast, long-distance networks. Based on these observations

Reviewer 2: @lisongxu Since we are using fast and long-distance networks in a few places, for consistency, we may use "fast and long-distance networks" instead of "fast, long-distance networks."

Author: done

test-bed experiments covering a wide range of network environments.

> More information can be found in {{HKLRX06}}.

Reviewer: ```suggestion
More information can be found in {{HKLRX06}}.
Additionally, there is decade-long deployment experience
with CUBIC on the Internet.
```

- initialize *W<sub>est</sub>* after timeout and remove variable
  *W<sub>last_max</sub>*
  ([#28](https://github.com/NTAP/rfc8312bis/issues/28))
  
- Rename TCP-Friendly to AIMD-Friendly and rename Standard TCP
  to AIMD TCP to avoid confusion as CUBIC has been widely used
  in the Internet.
  ([#38](https://github.com/NTAP/rfc8312bis/issues/38))

- Change introductory text to reflect the significant broader
  deployment of CUBIC in the Internet.
  ([#39](https://github.com/NTAP/rfc8312bis/issues/39))

>   

Reviewer: This should move to a new subsection called `## Since draft-eggert-tcpm-rfc8312bis-01` that should come before the section this text is currently in.

Author: done
